community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/combined/CombinedSchemaIndexProvider.java

```diff
@@ -0,0 +1,135 @@
+ */
+public class CombinedSchemaIndexProvider extends SchemaIndexProvider
+{
+    private final SchemaIndexProvider boostProvider;
+    private final SchemaIndexProvider fallbackProvider;
+
+    public CombinedSchemaIndexProvider( SchemaIndexProvider boostProvider, SchemaIndexProvider fallbackProvider )
+    {
+        super( new Descriptor( "combined", "0.1" ), 0 );
+        this.boostProvider = boostProvider;
+        this.fallbackProvider = fallbackProvider;
+    }
+
+    @Override
+    public IndexPopulator getPopulator( long indexId, IndexDescriptor descriptor, IndexSamplingConfig samplingConfig )
+    {
+        return new CombinedIndexPopulator(
+                boostProvider.getPopulator( indexId, descriptor, samplingConfig ),
+                fallbackProvider.getPopulator( indexId, descriptor, samplingConfig ) );
+    }
+
+    @Override
+    public IndexAccessor getOnlineAccessor( long indexId, IndexDescriptor descriptor,
+            IndexSamplingConfig samplingConfig ) throws IOException
+    {
+        return new CombinedIndexAccessor(
+                boostProvider.getOnlineAccessor( indexId, descriptor, samplingConfig ),
+                fallbackProvider.getOnlineAccessor( indexId, descriptor, samplingConfig ) );
+    }
+
+    @Override
+    public String getPopulationFailure( long indexId ) throws IllegalStateException
+    {
+        String boostFailure = boostProvider.getPopulationFailure( indexId );
+        String fallbackFailure = fallbackProvider.getPopulationFailure( indexId );
+        if ( boostFailure != null )
+        {
+            return fallbackFailure == null ? boostFailure : boostFailure + " and " + fallbackFailure;
+        }
+        else
+        {
+            return fallbackFailure;
+        }
+    }
+
+    @Override
+    public InternalIndexState getInitialState( long indexId, IndexDescriptor descriptor )
+    {
+        InternalIndexState boostState = boostProvider.getInitialState( indexId, descriptor );
+        InternalIndexState fallbackState = fallbackProvider.getInitialState( indexId, descriptor );
+        if ( boostState != fallbackState )
+        {
+            throw new IllegalStateException(
+                    format( "Internal providers answer with different state boost:%s, fallback:%s",
+                            boostState, fallbackState ) );
+        }
+        return boostState;
+    }
+
+    @Override
+    public StoreMigrationParticipant storeMigrationParticipant( FileSystemAbstraction fs, PageCache pageCache )
+    {
+        // TODO implementation of this depends on decisions around defaults and migration. Coming soon.
+        return StoreMigrationParticipant.NOT_PARTICIPATING;
+    }
+
+    static <T> T select( T boost, T fallback, Value... values )
+    {
+        if ( values.length > 1 )
+        {
+            // Multiple values must be handled by fallback
+            return fallback;
+        }
+
+        Value singleValue = values[0];
```

Comment:
We just talked about this but just want to type it out for completeness of this PR.

All index updates come as complete `IndexEntryUpdate` objects which contain the whole index entry that has been updated, not only the single part that is actually changed. So this should not be a problem.